apis/core/v1alpha2/readinessprovider_types.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ReadinessProviderState string
+
+const (
+	ProviderSuccessState    = ReadinessProviderState("success")
+	ProviderFailureState    = ReadinessProviderState("failure")
+	ProviderInProgressState = ReadinessProviderState("inprogress")
+)
+
+type ReadinessConditionState string
+
+const (
+	ConditionSuccessState    = ReadinessConditionState("success")
+	ConditionFailureState    = ReadinessConditionState("failure")
+	ConditionInProgressState = ReadinessConditionState("inprogress")
+)
+
+// ReadinessProviderSpec defines the desired state of ReadinessProvider
+type ReadinessProviderSpec struct {
+	// CheckRef is the name of the check that the current provider satisfies
+	CheckRef string `json:"checkRef"`
+
+	// Repeatable denotes if the provider should be re-evaluated periodically
+	//+kubebuilder:validation:Optional
+	Repeatable bool `json:"repeatable"`
+
+	// RepeatInterval is the re-evaluation interval
+	//+kubebuilder:validation:Optional
+	RepeatInterval *metav1.Time `json:"repeatInterval"`
+
+	// Conditions is the set of checks that must be evaluated to true to mark the provider as ready
+	Conditions []ReadinessProviderCondition `json:"conditions"`
+}
+
+type ReadinessProviderCondition struct {
+	// Name is the name of the condition
+	Name string `json:"name"`
+
+	// ResourceExistenceCondition is the condition that checks for the presence of a certain resource in the cluster
+	ResourceExistenceCondition ResourceExistenceCondition `json:"resourceExistenceCondition"`
+}
+
+type ResourceExistenceCondition struct {
+	// APIVersion is the API version of the resource that is being checked
+	APIVersion string `json:"apiVersion"`
+
+	// Kind is the API kind of the resource that is being checked
+	Kind string `json:"kind"`
+
+	// Namespace is the namespace of the resource that is being checked; if the Namespace is nil,
+	// the resource is assumed to be cluster scoped. Empty string for the namespace will throw error
+	//+kubebuilder:validation:Optional
+	Namespace *string `json:"namespace"`
+	Name      string  `json:"name"`
+}
+
+// ReadinessProviderStatus defines the observed state of ReadinessProvider
+type ReadinessProviderStatus struct {
+	// State is the computed state of the provider. The state will be success if all the coditions pass;
+	// The state will be failure if any of the conditions fail. Otherwise, the state will be in-progress
+	// +kubebuilder:validation:Enum=success;failure;inprogress
+	State ReadinessProviderState `json:"state"`
+
+	// Conditions is the set of ReadinessConditions that are being evaluated
+	Conditions []ReadinessConditionStatus `json:"conditions"`
+}
+
+type ReadinessConditionStatus struct {
```

Comment:
It may be useful to add a `Message string` property here, which can be populated with some info/debug/error messages

Reply:
Done